physical/redis.go

```diff
@@ -0,0 +1,314 @@
+package physical
+
+import (
+	"encoding/base64"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/garyburd/redigo/redis"
+)
+
+// RedisBackend is a physical backend that stores data at a specific prefix
+// within Redis.  This is not especially useful as an ordinary
+// secret-storage backend, unless you set up Redis persistence.
+//
+// But this backend offers a relatively simple way to set up a "medium
+// availability" Vault cluster, where you use S3 or another backend for
+// storage, and a single Redis server as a "ha_backend" to handle leader
+// elections.  If the Redis server fails, then all vault servers should go
+// to standby until Redis returns.
+//
+// This is potentially useful because Redis servers are ubiquitous and
+// surprisingly reliable, and because not everybody wants to set up and
+// administer either Consul or etcd for a single leader lock.  Also,
+// high-quality hosted and managed Redis services are widely available.
+type RedisBackend struct {
+	path   string
+	pool   *redis.Pool
+	logger *log.Logger
+}
+
+func newRedisBackend(conf map[string]string, logger *log.Logger) (Backend, error) {
+	// Get the Redis path from the configuration.
+	path, ok := conf["path"]
+	if !ok {
+		path = "vault"
+	}
+
+	// Get the Redis URL from our environment.
+	url := os.Getenv("REDIS_URL")
+	if url == "" {
+		url = "redis://127.0.0.1:6379"
+	}
+	addr := strings.TrimPrefix(url, "redis://")
+
+	// Create a Redis connection pool so that we can access Redis from
+	// multiple goroutines in parallel.
+	pool := &redis.Pool{
+		MaxIdle:     2,
+		IdleTimeout: 300 * time.Second,
+		Dial: func() (redis.Conn, error) {
+			conn, err := redis.Dial("tcp", addr)
+			// TODO: Add AUTH here if we want to support it.
+			return conn, err
+		},
+		TestOnBorrow: func(c redis.Conn, t time.Time) error {
+			_, err := c.Do("PING")
+			return err
+		},
+	}
+
+	return &RedisBackend{
+		path:   path,
+		pool:   pool,
+		logger: logger,
+	}, nil
+}
+
+func (c *RedisBackend) Put(entry *Entry) error {
+	conn := c.pool.Get()
+	defer conn.Close()
+
+	encoded := base64.StdEncoding.EncodeToString(entry.Value)
+	realKey := fmt.Sprintf("%s/%s", c.path, entry.Key)
+	_, err := conn.Do("SET", realKey, encoded)
+	return err
+}
+
+func (c *RedisBackend) Get(key string) (*Entry, error) {
+	conn := c.pool.Get()
+	defer conn.Close()
+
+	reply, err := conn.Do("GET", fmt.Sprintf("%s/%s", c.path, key))
+	if err != nil {
+		return nil, err
+	}
+	if reply == nil {
+		// No value at this key, so don't try to decode.
+		return nil, nil
+	}
+
+	replyStr, err := redis.String(reply, err)
+	if err != nil {
+		return nil, err
+	}
+
+	value, err := base64.StdEncoding.DecodeString(replyStr)
```

Comment:
i feel like encoding/decoding the values is unnecessary... i know the other backends do it, but i don't see its value in this context

just adds complexity